src/translate/axiom_rules.py

```diff
@@ -1,291 +1,138 @@
-        # heads of axioms]."
-        for init in list(axiom_init):
-            assert init.negate() in literals_in_heads
-        for literal in literals_in_heads:
-            assert ((literal.negated) == (literal.positive() in axiom_init)
-                    or (literal.negate() in literals_in_heads))
-
-    # 4. For every rule head <- ... cond ... where cond is a literal
-    #    of a derived variable where the layer of head is equal to
-    #    the layer of cond, cond occurs with the same polarity in heads.
-    #
-    # Note regarding issue454 and issue453: Because of the negated axioms
-    # mentioned in these issues, a derived variable may appear with *both*
-    # polarities in heads. This makes this test less strong than it would
-    # be otherwise. When these issues are addressed and axioms only occur
-    # with one polarity in heads, this test will remain correct in its
-    # current form, but it will be able to detect more violations of the
-    # layering property.
-    print("Verifying 4...")
-    for axiom in axioms:
-        head = axiom.effect
-        head_positive = head.positive()
-        body = axiom.condition
-        for cond in body:
-            cond_positive = cond.positive()
-            if (cond_positive in variables_in_heads and
-                axiom_layers[cond_positive] == axiom_layers[head_positive]):
-                assert cond in literals_in_heads
 
-    # 5. For every rule head <- ... cond ... where cond is a literal
-    #    of a derived variable, the layer of head is greater or equal
-    #    to the layer of cond.
-    print("Verifying 5...")
-    for axiom in axioms:
-        head = axiom.effect
-        head_positive = head.positive()
-        body = axiom.condition
-        for cond in body:
-            cond_positive = cond.positive()
-            if cond_positive in variables_in_heads:
-                # We need the assertion to be on a single line for
-                # our error handler to be able to print the line.
-                assert (axiom_layers[cond_positive] <= axiom_layers[head_positive]), (axiom_layers[cond_positive], axiom_layers[head_positive])
+def compute_necessary_literals(dependencies, goals, operators):
+    necessary_literals = set()
 
-def get_axioms_by_atom(axioms):
-    axioms_by_atom = {}
-    for axiom in axioms:
-        axioms_by_atom.setdefault(axiom.effect, []).append(axiom)
-    return axioms_by_atom
+    for g in goals:
+        if g.positive() in dependencies.derived_variables:
+            necessary_literals.add(g)
 
-def compute_axiom_layers(axioms, axiom_init):
-    # We include this assertion to make sure testing membership in
-    # axiom_init is efficient.
-    assert isinstance(axiom_init, set)
+    for op in operators:
+        derived_pres = (l for l in op.precondition if l.positive() in dependencies.derived_variables)
+        necessary_literals.update(derived_pres)
+
+        # TODO: why do we add both the literal and its negation here?
```

Comment:
I remember having a discussion with Malte about this but I don't understand anymore why we consider the negation of an effect condition as necessary. In my understanding we say literals are necessary if they *achieve* a necessary literal (or are a goal), but if a conditional effect does not fire, it does not achieve the negation of the effect.

Reply:
I am not sure I can fully answer this without looking again more carefully at what these "necessary_literals" are actually used for, i.e., what their effect is. Let me first try to explain this from a relevance analysis perspective, and ignoring derived variables altogether. That is, assume the task has no axioms, but it might have conditional effects. It might also have negative preconditions and negative goal conditions.

In relevance analysis, the "relevant literals" thing is all about the question: "can there ever be a reason why I might want this literal to be true?". If a literal is a goal, we might want it to be true.
=> goal literals are relevant

If a literal is an effect of an action that we want to be true, then this action looks useful, so might also want the preconditions of the action to be true.
=> preconditions of actions that achieve relevant literals are relevant

If a literal is a conditional effect of an action that we want to be true, then this conditional effect looks useful, so we might also want the effect conditions of this effect to be true
=> effect conditions of conditional effects that achieve relevant literals are relevant

Now that leaves the subtle case of unwanted conditional effects.
Look at a conditional effect of the form (A & not B => not C), and consider the case where C is relevant.
If the operator as a whole is relevant (i.e., it achieves a relevant literal -- not through this particular effect, but through another one), then a plan might include this operator. Because C is relevant it might include this operator in a context where we explicitly want this conditional effect to *not* trigger. And in order to achieve this, we might want to have "not A" or "B", because these prevent the conditional effect from triggering.

So this would give us a rule of the form
=> negations of effect conditions of conditional effects that achieve negations of relevant literals are relevant

(This rule ignores the side condition that we only need to do this if we have a reason to be interested in this operator in the first place, i.e., that we can first check that the operator has some relevant effect. Overapproximating this by dropping this condition doesn't make the rule wrong, but perhaps tighter if it needs to be.)

So this basically means that we would want rules of the form "add cond_lit if eff_lit is relevant" and "add cond_lit.negate() if eff_lit.negate() is relevant".

Note that "eff_lit is relevant" and "eff_lit.negate() is relevant" is not an either/or; both might be true, or neither.

I think these are essentially the same types of rules we also want for axioms (where cond_lit is a body literal and eff_lit is an effect literal). Of course, in axioms effect literals are always positive.

Hope this helps. I'm not 100% sure it does because I still need to understand again if this necessary literal computation here is the same kind of thing as a relevance analysis. But I think it might be.